File: src/Bus.cpp
#include "Bus.h"

Bus::Bus() {
  for (auto &i : ram)
    i = 0x00;
  cpu.ConnectBus(this);

  // Connect DMC memory read callback
  apu.SetReadCallback(
      [this](uint16_t addr) -> uint8_t { return this->read(addr, true); });
}

Bus::~Bus() {}

void Bus::insertCartridge(const std::shared_ptr<Cartridge> &cartridge) {
  this->cart = cartridge;
  ppu.ConnectCartridge(cartridge);
}

void Bus::reset() {
  cart->reset();
  cpu.reset();
  ppu.reset();
  apu.reset();
  nSystemClockCounter = 0;
}

void Bus::clock() {
  ppu.clock();

  if (nSystemClockCounter % 3 == 0) {
    // APU runs at CPU rate
    apu.clock();

    if (dma_transfer) {
      if (dma_dummy) {
        if (nSystemClockCounter % 2 == 1) {
          dma_dummy = false;
        }
      } else {
        if (nSystemClockCounter % 2 == 0) {
          dma_data = read(dma_page << 8 | dma_addr);
        } else {
          ppu.OAM[dma_addr] = dma_data;
          dma_addr++;

          if (dma_addr == 0x00) {
            dma_transfer = false;
            dma_dummy = true;
          }
        }
      }
    } else {
      cpu.clock();
    }

    // Check mapper IRQ (for MMC3) - Level Triggered
    // If IRQ is asserted by mapper, we signal the CPU.
    // The CPU handle the 'I' flag check and only interrupts on instruction
    // boundaries.
    if (cart->GetIRQState()) {
      cpu.irq();
    }
  }

  if (ppu.nmi) {
    ppu.nmi = false;
    cpu.nmi();
  }

  nSystemClockCounter++;
}

void Bus::write(uint16_t addr, uint8_t data) {
  if (cart->cpuWrite(addr, data)) {
  } else if (addr >= 0x0000 && addr <= 0x1FFF) {
    ram[addr & 0x07FF] = data;
  } else if (addr >= 0x2000 && addr <= 0x3FFF) {
    ppu.cpuWrite(addr & 0x0007, data);
  } else if (addr >= 0x4000 && addr <= 0x4013) {
    apu.cpuWrite(addr, data);
  } else if (addr == 0x4014) {
    dma_page = data;
    dma_addr = 0x00;
    dma_transfer = true;
  } else if (addr == 0x4015) {
    apu.cpuWrite(addr, data);
  } else if (addr >= 0x4016 && addr <= 0x4017) {
    if (addr == 0x4017) {
      apu.cpuWrite(addr, data);
    }
    controller_state[addr & 0x0001] = controller[addr & 0x0001];
  }
}

uint8_t Bus::read(uint16_t addr, bool bReadOnly) {
  uint8_t data = 0x00;
  if (cart->cpuRead(addr, data)) {
  } else if (addr >= 0x0000 && addr <= 0x1FFF) {
    data = ram[addr & 0x07FF];
  } else if (addr >= 0x2000 && addr <= 0x3FFF) {
    data = ppu.cpuRead(addr & 0x0007, bReadOnly);
  } else if (addr == 0x4015) {
    data = apu.cpuRead(addr);
  } else if (addr >= 0x4016 && addr <= 0x4017) {
    data = (controller_state[addr & 0x0001] & 0x80) > 0;
    controller_state[addr & 0x0001] <<= 1;
  }
  return data;
}

File: src/Mogiki.Core/Bus/Bus.cs
using System.Runtime.CompilerServices;
using Mogiki.Core.Apu;
using Mogiki.Core.Cartridge;
using Mogiki.Core.Cpu;
using Mogiki.Core.Ppu;

namespace Mogiki.Core.Bus;

/// <summary>
/// Main system bus interconnecting CPU, PPU, APU, RAM, Cartridge, and Controllers.
/// </summary>
public sealed class Bus
{
    public Cpu6502 Cpu { get; }
    public Ppu2C02 Ppu { get; }
    public Apu2A03 Apu { get; }
    public byte[] Ram { get; } = new byte[2048];
    public Cartridge.Cartridge? Cartridge { get; private set; }

    // Controllers
    public byte[] Controller { get; } = new byte[2];
    private readonly byte[] _controllerState = new byte[2];
    private bool _controllerStrobe;

    // Clock
    private uint _systemClockCounter;

    // DMA
    private byte _dmaPage;
    private byte _dmaAddr;
    private byte _dmaData;
    private bool _dmaTransfer;
    private bool _dmaDummy = true;

    public Bus()
    {
        Cpu = new Cpu6502 { Bus = this };
        Ppu = new Ppu2C02();
        Apu = new Apu2A03 { CpuReadCallback = addr => Read(addr, true) };
    }

    public void InsertCartridge(Cartridge.Cartridge cartridge)
    {
        Cartridge = cartridge;
        Ppu.Cartridge = cartridge;
    }

    public void RemoveCartridge()
    {
        Cartridge = null;
        Ppu.Cartridge = null;
    }

    public void Reset()
    {
        Cartridge?.Reset();
        Cpu.Reset();
        Ppu.Reset();
        Apu.Reset();
        _systemClockCounter = 0;
        _dmaTransfer = false;
        _dmaDummy = true;
        _controllerStrobe = false;
        _controllerState[0] = 0;
        _controllerState[1] = 0;
    }

    public void Clock()
    {
        Ppu.Clock();

        if ((_systemClockCounter % 3) == 0)
        {
            // APU runs at CPU rate
            Apu.Clock();

            if (_dmaTransfer)
            {
                if (_dmaDummy)
                {
                    if ((_systemClockCounter % 2) == 1)
                    {
                        _dmaDummy = false;
                    }
                }
                else
                {
                    if ((_systemClockCounter % 2) == 0)
                    {
                        _dmaData = Read((ushort)((_dmaPage << 8) | _dmaAddr));
                    }
                    else
                    {
                        Ppu.OAM[_dmaAddr] = _dmaData;
                        _dmaAddr++;

                        if (_dmaAddr == 0x00)
                        {
                            _dmaTransfer = false;
                            _dmaDummy = true;
                        }
                    }
                }
            }
            else
            {
                if (Cartridge != null && Cartridge.IrqState)
                {
                    Cpu.Irq();
                }
                Cpu.Clock();
            }
        }

        if (Ppu.Nmi)
        {
            Ppu.Nmi = false;
            Cpu.Nmi();
        }

        _systemClockCounter++;
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public double GetAudioSample() => Apu.GetOutputSample();

    public void Write(ushort addr, byte data)
    {
        Cartridge?.CpuSnoopWrite(addr, data);

        if (Cartridge != null && Cartridge.CpuWrite(addr, data))
        {
            return;
        }

        if (addr is >= 0x0000 and <= 0x1FFF)
        {
            Ram[addr & 0x07FF] = data;
        }
        else if (addr is >= 0x2000 and <= 0x3FFF)
        {
            Ppu.CpuWrite((ushort)(addr & 0x0007), data);
        }
        else if (addr is >= 0x4000 and <= 0x4013)
        {
            Apu.CpuWrite(addr, data);
        }
        else if (addr == 0x4014) // OAM DMA
        {
            _dmaPage = data;
            _dmaAddr = 0x00;
            _dmaTransfer = true;
        }
        else if (addr == 0x4015)
        {
            Apu.CpuWrite(addr, data);
        }
        else if (addr is >= 0x4016 and <= 0x4017)
        {
            if (addr == 0x4016)
            {
                bool strobe = (data & 0x01) != 0;

                // A high strobe continuously exposes the current A button.
                // The falling edge latches both controller ports for serial
                // reads, matching the NES controller protocol.
                if (strobe || _controllerStrobe)
                {
                    LatchControllers();
                }

                _controllerStrobe = strobe;
            }
            else
            {
                Apu.CpuWrite(addr, data);
            }
        }
    }

    public byte Read(ushort addr, bool readOnly = false)
    {
        if (Cartridge != null && addr is 0xFFFA or 0xFFFB)
        {
            Cartridge.CpuSnoopRead(addr);
        }

        if (Cartridge != null && Cartridge.CpuRead(addr, out byte data))
        {
            return data;
        }

        if (addr is >= 0x0000 and <= 0x1FFF)
        {
            return Ram[addr & 0x07FF];
        }

        if (addr is >= 0x2000 and <= 0x3FFF)
        {
            return Ppu.CpuRead((ushort)(addr & 0x0007), readOnly);
        }

        if (addr == 0x4015)
        {
            return Apu.CpuRead(addr);
        }

        if (addr is >= 0x4016 and <= 0x4017)
        {
            int port = addr & 0x0001;
            byte val = (byte)(((_controllerStrobe ? Controller[port] : _controllerState[port]) & 0x80) != 0 ? 1 : 0);

            if (!_controllerStrobe)
            {
                // Once all eight buttons have been shifted out, real NES
                // controllers keep returning 1 rather than 0.
                _controllerState[port] = (byte)((_controllerState[port] << 1) | 0x01);
            }

            return val;
        }

        return 0x00;
    }

    private void LatchControllers()
    {
        _controllerState[0] = Controller[0];
        _controllerState[1] = Controller[1];
    }
}
